Add duplicate functionality for unit, sub-section and section

mushtaqak opened the pull request:
# Description

This change adds duplicate functionality to section, subsection and unit on studio course outline.

### [TNL-5797](https://openedx.atlassian.net/browse/TNL-5797), [TNL-5798](https://openedx.atlassian.net/browse/TNL-5798)

### Sandbox
- [x] [Sandbox link](https://studio-duplicatexblock.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course)
### Testing
- [x] Unit
- [x] Jasmine
- [x] Manual
### Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @muzaffaryousaf 
- [x] Code review: @muhammad-ammar 
- [ ] A11y review: @cptvitamin  
- [ ] Doc: @catong  
- [x] Product: @sstack22 
### Post-review
- [x] Squash commits

mushtaqak commented:
jenkins run python

muzaffaryousaf commented:
@mushtaqak We might need a11y review too.

sstack22 commented:
This looks great! The one thing I noticed was that it seemed to take a long time for duplicate to finish. Not sure if that's just my connection right now, or if that's something we can improve in the future.

muzaffaryousaf commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -400,6 +400,69 @@ define(['jquery', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers', 'common/j
+                    EditHelpers.verifyNotificationShowing(notificationSpy, /Duplicating/);
+                    AjaxHelpers.respondWithError(requests);
+                    expect(getItemHeaders('section').length, 2);
+                    EditHelpers.verifyNotificationShowing(notificationSpy, /Duplicating/);
```
Don't we show any error message on failure ? I'd like to verify that stat as well.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -538,11 +538,22 @@ def _check_equality(self, source_usage_key, duplicate_usage_key, parent_usage_ke
+                return duplicated_item.display_name == original_item.category
+            return duplicated_item.display_name == original_item.display_name
+        if original_item.display_name is not None:
+            return duplicated_item.display_name == "Duplicate of '" + original_item.display_name + "'"
```
I would use `format` instead of `+`

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -538,11 +538,22 @@ def _check_equality(self, source_usage_key, duplicate_usage_key, parent_usage_ke
+            return duplicated_item.display_name == original_item.display_name
+        if original_item.display_name is not None:
+            return duplicated_item.display_name == "Duplicate of '" + original_item.display_name + "'"
+        return duplicated_item.display_name == "Duplicate of " + original_item.category
```
same like above.

muzaffaryousaf commented:
@mushtaqak I'm done with my first pass, let me know when it's ready for second review.

muhammad-ammar commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -653,6 +653,10 @@ def _duplicate_item(parent_usage_key, duplicate_source_usage_key, user, display_
                 duplicate_metadata[field.name] = field.read_from(source_item)
+
+        if is_child:
+            display_name = source_item.display_name if source_item.display_name is not None else source_item.category
```
@mushtaqak What about `display_name = source_item.display_name or source_item.category`?

mushtaqak commented on cms/static/js/views/pages/container.js:
```diff
@@ -226,22 +228,19 @@ define(['jquery', 'underscore', 'gettext', 'js/views/pages/base_page', 'common/j
-                            });
+                    parentElement = self.findXBlockElement(xblockElement.parent()),
+                    scrollOffset = ViewUtils.getScrollOffset(xblockElement),
+                    $placeholderEl = $(self.createPlaceholderElement()),
```
`eslint` quality was failing we can append element in the same expression.

sstack22 commented:
finished my review, 👍! Amazing work

mushtaqak commented:
@sstack22 Thanks :) May be you want to see the updated [sandbox](https://studio-duplicatexblock.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course) too. I have fixed delay, added focusable field for the duplicated block and it is expanded and scrolled down as well.

@muzaffaryousaf @muhammad-ammar [Sandbox](https://studio-duplicatexblock.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course) updated 2 hours ago :) Please review, I would like to get this merged tomorrow :)

muhammad-ammar commented:
👍

muzaffaryousaf commented on cms/static/js/views/pages/container.js:
```diff
@@ -209,10 +209,12 @@ define(['jquery', 'underscore', 'gettext', 'js/views/pages/base_page', 'common/j
                     listPanel = buttonPanel.prev(),
                     scrollOffset = ViewUtils.getScrollOffset(buttonPanel),
-                    placeholderElement = this.createPlaceholderElement().appendTo(listPanel),
+                    $placeholderEl = $(this.createPlaceholderElement()),
```
Would like to clean this with move work.

muzaffaryousaf commented:
👍
